### PR TITLE
 Fix #1875 - Nominator shouldn't ignore CONFIRM after externalization

### LINKED ISSUE #1875: Nominator should not ignore `CONFIRM` envelopes after externalization
This is happening now:
```
2021-04-01 10:24:18,388 Trace [agora.consensus.validation.Block] - idx 5 PublicKey: boa1xrdwry7vltf9mrzf62qgpdh282grqn9nlnvhzp0yt8y0y9zedmgh64s2qjg
2021-04-01 10:24:18,390 Trace [agora.consensus.validation.Block] - [source/agora/consensus/protocol/Nominator.d:907] idx #5: Validator boa1xrdwry7vltf9mrzf62qgpdh282grqn9nlnvhzp0yt8y0y9zedmgh64s2qjg has not yet signed block height 1
...
2021-04-01 10:24:18,391 Info [agora.node.Ledger] - Beginning externalization of block #1
2021-04-01 10:24:18,391 Info [agora.node.Ledger] - Transactions: 1 - Enrollments: 0
2021-04-01 10:24:18,391 Info [agora.node.Ledger] - Validators: Active: 6 - Signing: [248] - Slashed: []
...
2021-04-01 10:24:18,398 Trace [agora.consensus.protocol.Nominator] - Emitting envelope: { statement: { node: boa1xrdwry6f...z5en, slotIndex: 1, pledge: Externalize { commitQset: { hash: 0x1c66...d52c, quorum: <unknown> }, commit: { counter: 1, value: { tx_set: [0x0da1...0521], enrolls: [], missing_validators: [], time_offset: 1 } }, nh: 1 } }, sig: 0x5ef1...cd2d }
2021-04-01 10:24:18,399 Trace [agora.consensus.protocol.Nominator] - Received BLOCK SIG 0x0ab364e5c1364afab21c7aa8470a1544fc535be62da3bc4884120c1284841788 from node boa1xrdwryayr9r3nacx26vwe6ymy2kl7zp7dc03q5h6zk65vnu6mtkkzdqg39f for block 1
2021-04-01 10:24:18,399 Trace [agora.consensus.protocol.Nominator] - collectBlockSignature: [boa1xrdwryayr9r3nacx26vwe6ymy2kl7zp7dc03q5h6zk65vnu6mtkkzdqg39f] Enrollment commitment (CR): 0x30719c79055dd96c9a5ba718472ce5dd1577b64ce7b3a3d41629b572fd12f094, signing commitment (R): 0x942415dcb216843c6420916677a169661175585be70a499991489a5360e49b15
2021-04-01 10:24:18,400 Trace [agora.consensus.protocol.Nominator] - collectBlockSignature: VALID block signature at height 1 for node boa1xrdwryayr9r3nacx26vwe6ymy2kl7zp7dc03q5h6zk65vnu6mtkkzdqg39f
2021-04-01 10:24:18,400 Trace [agora.consensus.protocol.Nominator] - Updated block signatures for block 1, mask: [248]
2021-04-01 10:24:18,400 Trace [agora.consensus.protocol.Nominator] - Receiving envelope: { statement: { node: boa1xrdwry7v...2qjg, slotIndex: 1, pledge: Prepare { qset: { hash: 0x1c66...d52c, quorum: <unknown> }, ballot: { counter: 1, value: { tx_set: [0x0da1...0521], enrolls: [], missing_validators: [], time_offset: 1 } }, prep: { counter: 1, value: { tx_set: [0x0da1...0521], enrolls: [], missing_validators: [], time_offset: 1 } }, prepPrime: <null>, nc: 1, nH: 1 } }, sig: 0x7198...9e87 }
2021-04-01 10:24:18,400 Trace [agora.consensus.protocol.Nominator] - receiveEnvelope: Ignoring envelope with slot id cast(const(__c_ulonglong)) 1 as ledger is at height 1
2021-04-01 10:24:18,400 Trace [agora.consensus.protocol.Nominator] - Received BLOCK SIG 0x05d311e324a85050d177a16ba28475372321d982add38d7f67291f12fc8f5e16 from node boa1xrdwryuhc2tw2j97wqe3ahh37qnjya59n5etz88n9fvwyyt9jyvrvfq5ecp for block 1
```

So we externalize a block, and when we get the `CONFIRM` ballot later we drop it because we already updated the Ledger.
But the signature in it is a valid block signature and should be taken into account.

### PR DESCRIPTION
**Pull request by Geod24**
See each commit for details. I want to have a look at the log to confirm I haven't missed anything.

### CODE REVIEW

**Review by AndrejMitrovic — APPROVED** [review]
This looks sound to me.

**Review by hewison-chris — APPROVED** [review]
LGTM

File: source/agora/consensus/protocol/Nominator.d [hunk 1]
```diff
@@ -512,6 +518,21 @@ extern(D):
...
+
+            // If it's an old envelope, we're only interested in the signature
+            if (envelope.statement.slotIndex <= last_block.header.height)
+            {
+                const sig = ValidatorBlockSig(
+                    Height(envelope.statement.slotIndex), public_key,
+                    Scalar(envelope.statement.pledges.confirm_.value_sig));
+                auto blocks = this.ledger.getBlocksFrom(Height(envelope.statement.slotIndex));
+                if (!this.collectBlockSignature(sig, blocks[0].hashFull()))
+                    log.trace("Added signature for {} from CONFIRM ballot", public_key);
```
Comment (linked0) [inline source/agora/consensus/protocol/Nominator.d:520]: Is this the right message?

Comment (Geod24) [reply source/agora/consensus/protocol/Nominator.d:520]: I think so, what do you see as problematic ?

Comment (linked0) [reply source/agora/consensus/protocol/Nominator.d:520]: It returns `false`, which means `Invalid point from public key` or `INVALID Block signature received`.

Comment (Geod24) [reply source/agora/consensus/protocol/Nominator.d:520]: 🤦 
You're right, I'll fix

**Review by linked0 — APPROVED** [review]
LGTM

File: source/agora/consensus/protocol/Nominator.d [hunk 2]
```diff
@@ -456,8 +456,14 @@ extern(D):
...
             return;
 
+        // Outdated envelopes might contain block signatures which we want to add
+        // Hence, define a certain tolerance for us to accept those.
+        // See https://github.com/bosagora/agora/issues/1875
+        static immutable ConfirmTolerance = 10;
+
         const Block last_block = this.ledger.getLastBlock();
-        if (Height(envelope.statement.slotIndex) != last_block.header.height + 1)
+        // Don't use `height - tolerance` as it could undeflow
```
Comment (kchulj) [inline source/agora/consensus/protocol/Nominator.d]: typo

**Review by kchulj — APPROVED** [review]
LGTM
one typo

**Comment by Geod24** [comment]
Broken down into two smaller PR, closing.

**Pull request closed without merging**